lib/util.js

```diff
@@ -151,6 +153,59 @@ function fetch_html(uri, encoding) {
     });
 }
 
+async function _fetch_article(hatch, uri, encoding, Article) {
+    const asset = new Article();
+    let $;
+    try {
+        $ = await fetch_html(uri, encoding);
+    } catch (error) {
+        if (error.message.includes('non-successful status code')) {
+            logger.error(error);
+            asset.set_has_failed();
```

Comment:
Instead of setting this flag here on the asset, couldn't we directly add the asset to the failed list in the hatch? We could remove the awkward `set_has_failed()` API

Reply:
Yes this `set_has_failed()` API is awkward. It is just for checking in the ingester:
```
        if (asset.has_failed()) {
            return;
        }
```
But I can add a new API for the Hatch, `Hatch.has_failed_asset(asset)` instead.